projects/control-service/projects/gpu_management/README.md

```diff
@@ -0,0 +1,103 @@
+# A project for Efficient Management of GPUs
+
+
+Initial setup.
+1. Install postgres if not already installed
+2. Update the application.properties with the postgres connection details for this instance.
+3. Download the metabase oss jar from here: https://www.metabase.com/docs/latest/installation-and-operation/running-the-metabase-jar-file
+4. startup metabase with `java -jar metabase.jar`
+5. login with gpuManager@gpus.com and pw: gpu123
+6. click settings in top right corner, choose admin setting
+7. Choose databases from the top row, choose 'demo' update the credentials to be your local postgres database
+run java -jar
+
+
+
+## Algorithm Implementation
+
+
+
+
+### Reshuffle
+To reshuffle the jobs we use a linear programming solver.
+Linear programming is an optimization technique for a system of linear constraints and a linear objective function.
+
+Before you can use linear programming we need to decide what the variables are for our system.
+In our case we will create a list of variables where each one represents if a node is on that machine.
+1 => the job is running on the node. 0 => job is not running on the node
+For example if there are 6 jobs and 3 machines our variables will look like.
+
+|      | Job 1 | Job 2 | Job 3 | Job 4 | Job 5 | Job 6 |
+|-----------|-------|----|----|---|---|---|
+| Machine 1 | ✅(1)  | ❌(0) | ✅(1) |  ❌(0)  |  ❌(0)  |  ❌(0)  |
+| Machine 2 | ❌(0)  | ❌(0) |  ❌(0)| ✅(1) |  ❌(0)  |  ❌(0)  |
+| Machine 3 | ❌(0)  | ❌(0) | ❌(0) | ❌(0) | ✅(1) |  ❌(0)  |
+
+
+In this case we have 18 (jobs*nodes) variables.
+
+
+Next we need to decide on the constraints of our system.
+
+
+#### Constraint 1:
+A job can only exist on a single machine.
+It would be senseless to deploy a job to more than one machine.
+
+|                               | Job 1 | Job 2 | Job 3 | Job 4 | Job 5 | Job 6 |
+|-------------------------------|--|--|--|--|--|--|
+| Machine 1                     |  ✅(1) | ❌(0) |  ✅(1) | ❌(0) | ❌(0) | ❌(0) |
+| Machine 2                     | ❌(0) | ❌(0) | ❌(0) |  ✅(1) | ❌(0) | ❌(0) |
+| Machine 3                     | ❌(0) | ❌(0) | ❌(0) | ❌(0) |  ✅(1) | ❌(0) |
+| Constaint 1:<br/> Column <= 1 |   1 |  0 | 1 | 1 | 0 | 0 |
+
+```
+job_1___machine_1 + job_1___machine_2 + job_1___machine_3 <= 1
+job_2___machine_1 + job_2___machine_2 + job_2___machine_3 <= 1
+job_3___machine_1 + job_3___machine_2 + job_3___machine_3 <= 1
+job_4___machine_1 + job_4___machine_2 + job_4___machine_3 <= 1
+job_5___machine_1 + job_5___machine_2 + job_5___machine_3 <= 1
+job_6___machine_1 + job_6___machine_2 + job_6___machine_3 <= 1
+```
+
+
+#### Constraint 2
+The total number of jobs on a machine must be less than the amount of resources available on that machine.
```

Comment:
With GPUs limits==requests==actual_used always. 
This is because GPUs need to be provisioned at pod start time and can't be shared between pods. 

So we only need to keep track of what jobs are running successfully and how much resources we specified for them. 


It is actually the fact that limits and requests must be the same for GPUs that introduces the need for a system like that.

Reply:
> limits==requests==actual_used always

That seems wrong. 

Yes, I know that sharing GPU resources efficiently between different pods is currently impractical. 
Yes, I understand (learned now) it's currently not possible to add/remove dynamically GPUs to pods so limits==requests make sense.

But the usage is not the same. A job could be using all, some, or nearly none of the GPU's capacity at any given time, depending on its workload. I can see there are system to track usage like https://github.com/NVIDIA/dcgm-exporter

Reply:
It doesn't really matter how much of the device they are using no one else can use the rest of it. 

So if they say they want one device. Even if they are only using it for a few seconds every hour. By default no one else can use any of it. 

If you want we can hop on a call and discuss!